src/sst/core/model/python/pymodel_unitalgebra.cc

```diff
@@ -0,0 +1,389 @@
+//
+// Copyright (c) 2009-2020, NTESS
+// All rights reserved.
+//
+// This file is part of the SST software package. For license
+// information, see the LICENSE file in the top level directory of the
+// distribution.
+
+#include "sst_config.h"
+#include "sst/core/warnmacros.h"
+
+DISABLE_WARN_DEPRECATED_REGISTER
+#include <Python.h>
+REENABLE_WARNING
+
+#include <string.h>
+#include <sstream>
+
+#include <sst/core/warnmacros.h>
+#include <sst/core/model/python/pymacros.h>
+#include <sst/core/model/python/pymodel.h>
+#include <sst/core/model/python/pymodel_unitalgebra.h>
+
+#include <sst/core/unitAlgebra.h>
+
+using namespace SST::Core;
+extern SST::Core::SSTPythonModelDefinition *gModel;
+
+extern "C" {
+
+
+static int unitAlgebraInit(UnitAlgebraPy_t *self, PyObject *args, PyObject *UNUSED(kwds))
+{
+    char *init_str = NULL;
+    // PyObject* obj;
+    UnitAlgebraPy_t* obj;
+    
+    
+    if ( PyArg_ParseTuple(args, "s", &init_str) ) {
+        self->obj = new UnitAlgebra(init_str);
+        return 0;
+    }
+    PyErr_Clear();
+    if ( PyArg_ParseTuple(args, "O!", &PyModel_UnitAlgebraType, &obj) ) {
+        self->obj = new UnitAlgebra(*(obj->obj));
+        return 0;
+    }
+    PyErr_SetString(PyExc_TypeError,"sst.UnitAlgebra can only be initialized with another sst.UnitAlgebra or with a formatted string");
+    return -1;
+}
+
+static void unitAlgebraDealloc(UnitAlgebraPy_t *self)
+{
+    if ( self->obj ) delete self->obj;
+    Py_TYPE(self)->tp_free((PyObject*)self);
+}
+    
+static PyObject* unitAlgebraStr(PyObject* self)
+{
+    UnitAlgebraPy_t *ua = (UnitAlgebraPy_t*)self;
+    return SST_ConvertToPythonString(ua->obj->toStringBestSI().c_str());
+}
+    
+static PyObject* unitAlgebraRichCmp (PyObject *self, PyObject *other, int op)
+{
+    UnitAlgebraPy_t *self_ua = (UnitAlgebraPy_t*)self;
+    UnitAlgebraPy_t *other_ua = (UnitAlgebraPy_t*)other;
+    switch (op) {
+    case Py_LT:
+        if ( *(self_ua->obj) < *(other_ua->obj) ) return Py_True;
+        else return Py_False;
+        break;
+    case Py_LE:
+        if ( *(self_ua->obj) <= *(other_ua->obj) ) return Py_True;
+        else return Py_False;
+        break;
+    case Py_GT:
+        if ( *(self_ua->obj) > *(other_ua->obj) ) return Py_True;
+        else return Py_False;
+        break;
+    case Py_GE:
+        if ( *(self_ua->obj) >= *(other_ua->obj) ) return Py_True;
+        else return Py_False;
+        break;
+    case Py_EQ:
+    case Py_NE:
+    default:
+        return Py_NotImplemented;
+        break;
+    }
+    return Py_NotImplemented;
+}
+
+
+// Numerical functions
+static PyObject* unitAlgebraAddIP(PyObject *self, PyObject *other)
+{
+    UnitAlgebraPy_t *self_ua = (UnitAlgebraPy_t*)self;
+    UnitAlgebraPy_t *other_ua = (UnitAlgebraPy_t*)other;
+    *(self_ua->obj) += *(other_ua->obj);
+
+    PyObject *argList = Py_BuildValue("(O)", self);
+    PyObject* ret = PyObject_CallObject((PyObject*)&PyModel_UnitAlgebraType, argList);    
```

Comment:
This was pretty confusing to me at first about why we have to return a new reference. I would probably add a comment here about why this is being done.  This also gets repeated in several places.  Probably better to just make a function `returnNewInPlaceReference` or something.

Reply:
In turns out that with the semantics of python assignment and in-place operators, the in-place and regular math operations ended up being the same code.  I added a comment in the Number methods declaration to explain that.